### DISCUSSION TITLE
How does the Dockerfile work, and how could I use it myself?

### DISCUSSION THREAD

**Discussion by rafaeltab** [Kurabu-chan/Kurabu#66] [Q&A]
# How does it work?
Each application that should be built using docker has three things:
* A Dockerfile
* Three scripts
    * build:docker
    * build:docker-dev
    * build:docker-prod
* A service inside the kubernetes configuration

## The Dockerfile
Take as an example the [Dockerfile for the api](https://github.com/Kurabu-chan/Kurabu/blob/main/apps/api/docker/Dockerfile)
It is completely compatible with multiple architectures.

The most important steps are:
* prune all non-depended-upon workspaces, specific for the app.
* Install dependencies
* Build the application
* Start the application

## Using the Dockerfile
For building the docker image it is imperitive that the build is ran from the root of the monorepo.
However, this would disallow using the package.json of the specific app.
Luckily, docker allows specifying the location a build should be run from, you could even put the dockerfile on a different drive than the actual source code.

Currently, a [script](https://github.com/Kurabu-chan/Kurabu/blob/main/apps/api/scripts/docker.ts) is used to construct the actual commands used for building the image.
The main reason the script is used is because the build steps for the different environments are very similar, however, it requires multiple commands to run (multiple architectures require a custom buildx driver).

In the core the command is as follows: `docker buildx build -t <the tag to give the image> -f <path to docker_file> ../../`

## Kubernetes
If anyone wants an explanation of how the Kubernetes works, please leave a comment, and I might consider creating a post on the Kubernetes side of things.

# FAQ

## The Dockerfile doesn't seem to work, and I am using npm, yarn v2/v3 or pnpm
Sadly, it seems that turborepo is only compatible with yarn v1 for turbo prune at the moment. 
It could be that by now support for yarn v2 and v3 has come out, and also for pnpm, I have not tested this personally. 
**However, npm is not supported.**

**Comment by anthonyshew** [comment]
Perhaps a comment here https://github.com/Kurabu-chan/Kurabu/blob/46de0f4263d2631a5490fc6ea45fee6c23a03357/apps/api/docker/Dockerfile#L12 about why you're copying everything instead of only the packages needed for this app. (The trade off on modularity vs. performance for copying less files.)

**Reply by rafaeltab** [reply]
https://github.com/Kurabu-chan/Kurabu/pull/68
https://github.com/Kurabu-chan/Kurabu/issues/67
